server/src/main/java/org/candlepin/controller/CandlepinPoolManager.java

```diff
@@ -1227,36 +1232,163 @@ void removeEntitlement(Entitlement entitlement,
         sink.queueEvent(event);
     }
 
+
+    /**
+     * Revokes the given set of entitlements.
+     *
+     * @param entsToRevoke entitlements to revoke
+     */
     @Override
     @Transactional
-    public void revokeEntitlement(Entitlement entitlement) {
-        removeEntitlement(entitlement, true);
+    public void revokeEntitlements(List<Entitlement> entsToRevoke) {
+        if (entsToRevoke.isEmpty()) {
+            return;
+        }
+
+        List<Pool> poolsToLock = new ArrayList<Pool>();
+        for (Entitlement ent: entsToRevoke) {
+            poolsToLock.add(ent.getPool());
+        }
+
+        poolCurator.lock(poolsToLock);
+
+        log.info("Batch revoking entitlements: " + entsToRevoke.size());
+
+        entsToRevoke =  new ArrayList<Entitlement>(entsToRevoke);
+
+        List<Pool> poolsToDelete = poolCurator.listBySourceEntitlements(entsToRevoke);
+
+        for (Pool pool : poolsToDelete) {
+            entsToRevoke.addAll(pool.getEntitlements());
+        }
+
+        poolCurator.lock(poolsToDelete);
+
+        for (Entitlement ent : entsToRevoke) {
+            //We need to trigger lazy load of provided products
+            //to have access to those products later in this method.
+            ent.getPool().getProvidedProducts().size();
+            Pool pool = ent.getPool();
+            pool.setConsumed(pool.getConsumed() - ent.getQuantity());
+
```

Comment:
This is a great thing for quite a long discussion. Right now I don't think there is a problem with missing entitlements.remove.

However...

In this branch, the code that you mention is in EntitlementCurator.batchDelete. I don't think it should be there (it causes performance issues when Pool has a lot of entitlements).

If we never initialize pool.entitlements during revocation, then we won't run into any issues. If we do initialize pool.entitlements at some point and then do not call pool.getEntitlements().remove(ent); then the ent will stay in the database. 

I dont think the old code to which you refer to ever worked, because entitlements are EXTRA lazy, the actual remove did not initialize the collection and didn't remove anything from database. Later if you initialized the collection, the ent would stay in the database regardless.